signature/src/lib.rs

```diff
@@ -24,9 +45,481 @@ pub extern "C" fn verify_rsa(
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            VALID_TIMESTAMP,
+            ROOT_HASH,
+            "example.com",
+        );
+        assert_eq!(r, Err(BalrogError::HostnameMismatch));
+    }
+
+    #[test]
+    fn test_verify_fails_if_cert_has_expired() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1215559719, // July 9, 2008
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        assert_eq!(r, Err(BalrogError::CertificateExpired));
+    }
+
+    #[test]
+    fn test_verify_fails_if_chain_out_of_order() {
+        // Parse the chain then swap the leaf and intermediate certs.
+        let mut pem: Vec<Pem> = match parse_pem_chain(VALID_CERT_CHAIN) {
+            Err(e) => panic!("Found unexpected error: {}", e),
+            Ok(x) => x,
+        };
+        let leaf = pem.remove(0);
+        pem.insert(1, leaf);
+
+        let b = match Balrog::new(&pem) {
+            Err(e) => panic!("Found unexpected error: {}", e),
+            Ok(x) => x,
+        };
+        let r = b.verify_chain(VALID_TIMESTAMP, ROOT_HASH);
+        assert_eq!(r, Err(BalrogError::ChainSubjectMismatch));
+    }
+
+    #[test]
+    fn test_verify_fails_if_bad_chain_signature() {
+        let mut pem: Vec<Pem> = match parse_pem_chain(VALID_CERT_CHAIN) {
+            Err(e) => panic!("Found unexpected error: {}", e),
+            Ok(x) => x,
+        };
+        // An ASN.1 X509 certificate will take the form of:
+        // Certificate ::= SEQUENCE {
+        //     tbsCertificate     TBSCertificate,
+        //     signatureAlgorithm AlgorithmIdentifier,
+        //     signature          BITSTRING
+        // }
+        //
+        // This means the signature will always reside in the trailing bytes of
+        // the DER-encoded certificate, and we should be able to corrupt it by
+        // flipping a few bits.
+        let tail = pem[1].contents.pop().unwrap();
+        pem[1].contents.push(tail ^ 0x80);
+
+        let b = match Balrog::new(&pem) {
+            Err(e) => panic!("Found unexpected error: {}", e),
+            Ok(x) => x,
+        };
+        let r = b.verify_chain(VALID_TIMESTAMP, ROOT_HASH);
+        assert_eq!(
+            r,
+            Err(BalrogError::from(X509Error::SignatureVerificationError))
+        );
+    }
+
+    #[test]
+    fn test_verify_fails_if_bad_content_signature() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            &VALID_SIGNATURE.to_ascii_lowercase(), // altering case should modify the base64 signature without changing its length.
+            VALID_TIMESTAMP,
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        assert_eq!(
+            r,
+            Err(BalrogError::from(X509Error::SignatureVerificationError))
+        );
+    }
+
```

Comment:
I think we are missing tests for invalid input.

Reply:
Should be added now.